### ISSUE TITLE
Dynamic paging support (or a paging example)

### ISSUE THREAD

**Issue by designerx2** [segmentio/nightmare#424]
Is dynamic paging possible with Nightmare? By dynamic paging I mean a sort of "do while" function that will continue to goto and evaluate pages. 

I'm attempting to scrape ajax pages which require a bit of setup. It appears the evaluate function doesn't allow javascript functions to be passed in as arguments and I'm stuck. 

Any potential assistance is greatly appreciated. My apologies if this is the wrong spot for such inquiries. 

TIA

Nightmare and Electron appear to be a promising modern alternative to PhantomJS with CasperJS. The simplicity of the examples with generators as opposed to a mess of callbacks is very appealing. Kudos to the team!

**Comment by designerx2** [comment]
Below is a simple example of returning the links found on page 1 of a google search. For examples sake I'd like to return the links of all result pages (as of this writing there were 13 result pages.) 

```javascript
var Nightmare = require('nightmare');
var vo = require('vo');

vo(function* () {
  var nightmare = Nightmare({ show: true });
  var links = yield nightmare
    .goto('http://www.google.com')
    .wait('input[title="Search"]')
    .click('input[title="Search"]')
    .type('input[title="Search"]', 'Anequim Project')
    .click('input[name="btnK"]')
    .wait(600)
    .evaluate(function(){
        var linkArray = [];
        var links = document.querySelectorAll('h3.r a');
        for (var i = 0; i < links.length; ++i) {
            linkArray.push(links[i].getAttribute('href'));
        }
        return linkArray;
    });
  yield nightmare.end();
  return links;
})(function (err, result) {
  if (err) return console.log(err);
  console.log(result);
});

**Comment by rosshinkley** [comment]
You also might want to have a look at [my comment](https://github.com/segmentio/nightmare/issues/402#issuecomment-166948532) in #402.  It seems to be pretty close to what you're asking for.

**Comment by designerx2** [comment]
@rosshinkley ,  Yes.. that is very helpful. thank you for pointing it out. It appears multiple instances of Nightmare are fired up, I'm not sure I understand how Nightmare retains it's page scope yet fire up multiple instances. Oddly it is currently returning results of only 12 of the 14 result pages for me at this point.   I attempted searching a different term with greater than 100 page results and was presented with only 33 links, I haven't identified the problem yet. To resolve the EventEmitter errors I used the fix mentioned here: EventEmitter memory leak (https://github.com/segmentio/nightmare/issues/350).  

```javascript
var Nightmare = require('nightmare');
var vo = require('vo');

vo(run)(function(err, result) {
    if (err) throw err;
});

function* run() {
    var nightmare = Nightmare({ show: true }), 
        MAX_PAGE = 100, 
        currentPage = 0, 
        nextExists = true, 
        links = []; 

    yield nightmare 
        .goto('http://www.google.com')
        .wait('input[title="Search"]')
        .click('input[title="Search"]')
        .type('input[title="Search"]', 'Anequim Project')
        .click('input[name="btnK"]') 
        .wait(2000)

    nextExists = yield nightmare.visible('#pnnext'); 

    while (nextExists && currentPage < MAX_PAGE) { 
        links.push(yield nightmare 
            .evaluate(function() { 
                var linkArray = [];
                var links = document.querySelectorAll('h3.r a');
                return links[0].href; 
            })); 

        yield nightmare 
            .click('#pnnext')
            .wait(2000)

        currentPage++; 
        nextExists = yield nightmare.visible('#pnnext'); 
    } 

    console.dir(links); 
    yield nightmare.end(); 
} 
```

**Comment by rosshinkley** [comment]
@designerx2 I think there are two problems here.  The first is one I should have thought of and missed in my haste.  The example in #402 - and the one you're showing here - is naive and will miss the last page (if there is a "last" page) as the `#pnnext` button won't exist.  The second has to do with how Google shows pages.  I added a `scrollTo` to look at the pages before , and at page 13, the results fall below the omission threshold.  A picture might be helpful:

![test](https://cloud.githubusercontent.com/assets/1823131/12127088/36b0d5e4-b3c3-11e5-9961-74172bf6b8e6.png)

In short: there are only 13 result pages and the last one is being omitted.

**Comment by YueShuXie** [comment]
Wow, It's cool!

**Comment by rosshinkley** [comment]
I'm considering this issue resolved.  If you're still having this problem, feel free to reopen/open a new issue.
